### PR TITLE
Add Balanced Classification Data Splitter

### LINKED ISSUE #973: Balanced sampling for classification problems
#971 tracks adding a data check to detect and warn about severe class-imbalance. #457 tracks adding an automl objective which can perform well on imbalanced datasets

Detecting and warning about class imbalance could be helpful, but ideally we could improve our classifier quality by down-sampling the training data to flatten the distribution of the target over each class.

There's a couple levels this problem exists at:
1. In automl, if we want to make a fair comparison between models, we should either use an automl objective which is hardened to imbalanced data (like Neyman-Pearson tracked by #457), or we should balance the data before computing the objective score.
2. During the training of each pipeline, each CV fold should be down-sampled to have a balanced distribution of labels. A simple way to do this is to compute the splits in automl as we do now, apply down-sampling to balance, and then train/evaluate each pipeline on each split. We could also have the pipelines generate the splits and down-sample internally, but if this is done right there wouldn't be any difference in accuracy.

I see [sklearn's doc mentions class imbalance](https://scikit-learn.org/stable/modules/cross_validation.html#cross-validation-iterators-with-stratification-based-on-class-labels), but I don't agree with their recommendation "to ensure that relative class frequencies is approximately preserved in each train and validation fold." I suppose we could put this to the test and see which strategy seems to generalize more effectively.

### PR DESCRIPTION
**Pull request by bchen1116**
fix #973 

Perf test results [here](https://alteryx.quip.com/AXqlAwW8CV3w/Balanced-Classification-Data-Splitter-perf-tests)
Updated and more in-depth perf test [here](https://alteryx.quip.com/hSg1A2sTAq4n/Balanced-Classification-data-splitter-follow-up-tests)
Summary:
- The data sampler performs slightly worse (in holdout log loss, holdout f1, and time to fit) compared to the original `StratifiedKfold`
- This performance might be due to our default values for `min_samples/min_percentage`, so might be useful to do tuning.

### CODE REVIEW

File: evalml/tests/preprocessing_tests/test_balanced_classification_data_splitter.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    if dataset == 0:
+        X, y = X_y_binary
+    else:
+        X, y = X_y_multi
+    # make imbalanced
+    X_extended = np.append(X, X, 0)
+    y_extended = np.append(y, np.array([0] * len(y)), 0)
+    skf = StratifiedKFold(shuffle=True, n_splits=3, random_state=0)
+    sample_method = BalancedClassificationSampler(balanced_ratio=1, min_samples=50, random_seed=0)
+    data_splitter = BalancedClassificationDataCVSplit(balanced_ratio=1, min_samples=50, shuffle=True, n_splits=3, random_seed=0)
```
Comment (chukarsten) [inline evalml/tests/preprocessing_tests/test_balanced_classification_data_splitter.py:88]: Do you think this is worth maybe parameterizing this test to deal with both BalancedClassificationDataCVSplit and BalancedClassificationDataCVSplit?  I know there's a way you can package parameters together as you'd need to bundle like skf, sample_method and data_splitter together, but it seems the rest is the same.

**Review by chukarsten — CHANGES_REQUESTED** [review]
Solid work and it looks like it's really coming along.  This is going to add a lot of good functionality to the baseline.

I think a few things that can definitely change are parametrizing the test, replacing two code chunks with the relevant function in the parent class, modifying the transform name/docstring.  The other few things I'm not sure about.

Again, fantastic work here and thanks for bearing with through the requests...I know it's tiresome looking through the same PRs over and over.

File: evalml/preprocessing/data_splitters/balanced_classification_splitter.py [hunk 2]
```diff
@@ -89,4 +95,66 @@ def fit_resample(self, X, y):
...
+        y_ww = infer_feature_types(y)
+        X = _convert_woodwork_types_wrapper(X_ww.to_dataframe())
+        y = _convert_woodwork_types_wrapper(y_ww.to_series())
+        index_df = pd.Series(y.index)
+        for train, test in self.splitter.split(X, y):
+            X_train, y_train = X.iloc[train], y.iloc[train]
+            train_index_drop = self.sampler.fit_resample(X_train, y_train)
+            # convert the indices of the y column into index indices of the original pre-split y
+            train_indices = index_df[index_df.isin(train_index_drop)].dropna().index.values.tolist()
+            yield iter([train_indices, test])
```
[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/preprocessing/data_splitters/balanced_classification_splitter.py:160]: @chukarsten For CV cases, since split should be an iterable (over the number of splits), we want to return an iterator for the user so they can call `for train, test in s.split(X, y)`. If we `return` instead, we'll only get 1 split

File: evalml/tests/automl_tests/test_automl_search_classification.py [hunk 3]
```diff
@@ -84,6 +88,13 @@ def test_data_splitter(X_y_binary):
     assert isinstance(automl.rankings, pd.DataFrame)
     assert len(automl.results['pipeline_results'][0]["cv_data"]) == cv_folds
 
+    automl = AutoMLSearch(X_train=X, y_train=y, problem_type='binary', data_splitter=BalancedClassificationDataCVSplit(n_splits=cv_folds), max_iterations=1,
```
[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/tests/automl_tests/test_automl_search_classification.py:91]: @chukarsten I left it since it would be nice to know we still support `StratifiedKFold` as a sanity check, but I can remove it if it could be problematic/unnecessary?

File: evalml/tests/automl_tests/test_automl_utils.py [hunk 4]
```diff
@@ -94,7 +95,10 @@ def test_make_data_splitter_parameters(problem_type, expected_data_splitter):
     assert isinstance(data_splitter, expected_data_splitter)
     assert data_splitter.n_splits == 5
     assert data_splitter.shuffle
-    assert data_splitter.random_state == random_seed
+    if str(problem_type) == 'regression':
+        assert data_splitter.random_state == random_seed
```
[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/tests/automl_tests/test_automl_utils.py:99]: Right, for regression cases, we still use sklearn's `KFold` splitter, which is why we need to check random state here. If we want to encompass this with our own data splitter, then we'd be able to use `assert data_splitter.random_seed == random_seed`. I'm not aware of any current issues to track this, but is this something we want to change?

File: evalml/tests/preprocessing_tests/test_balanced_classification_data_splitter.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    BalancedClassificationDataTVSplit,
+    BalancedClassificationSampler,
+    TrainingValidationSplit
+)
+
+
+@pytest.mark.parametrize("splitter",
+                         [BalancedClassificationDataCVSplit,
+                          BalancedClassificationDataTVSplit])
+def test_data_splitter_nsplits(splitter):
```
[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/tests/preprocessing_tests/test_balanced_classification_data_splitter.py:17]: @chukarsten  I'm not sure I understand the question here

File: evalml/tests/preprocessing_tests/test_balanced_classification_data_splitter.py (hunk 1 above)
Comment (bchen1116) [reply evalml/tests/preprocessing_tests/test_balanced_classification_data_splitter.py:88]: Yep, good idea!

**Review by chukarsten — APPROVED** [review]
image

File: evalml/tests/preprocessing_tests/test_balanced_classification_data_splitter.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/tests/preprocessing_tests/test_balanced_classification_data_splitter.py:17]: Ah, I did have that hardcoded before for TV splits. I've changed it to just rely on the `n_splits` arg now.

File: evalml/data_checks/class_imbalance_data_check.py [hunk 6]
```diff
@@ -99,7 +99,7 @@ def validate(self, X, y):
         if len(below_threshold) and len(sample_counts):
             sample_count_values = sample_counts.index.tolist()
             severe_imbalance = [v for v in sample_count_values if v in below_threshold]
-            warning_msg = "The following labels have severe class imbalance because they fall under {:.0f}% of the target and have less than {} samples: {}"
+            warning_msg = "The following labels in the target have severe class imbalance because they fall under {:.0f}% of the target and have less than {} samples: {}"
```
Comment (bchen1116) [inline evalml/data_checks/class_imbalance_data_check.py:102]: Address comment Dylan left in the class imbalance data check PR

File: evalml/automl/automl_search.py [hunk 7]
```diff
@@ -600,6 +600,10 @@ def _find_best_pipeline(self):
                     X_train, X_threshold_tuning, y_train, y_threshold_tuning = split_data(X_train, y_train, self.problem_type,
                                                                                           test_size=0.2,
                                                                                           random_seed=self.random_seed)
+                if hasattr(self.data_splitter, "transform_sample"):
```
Comment (freddyaboulton) [inline evalml/automl/automl_search.py:603]: I'm not a big fan of using `hasattr` here. Ideally, the logic of knowing when and how to resample the data should live in one place. Right now it's in `BaseUnderSamplingSplitter`, here, and soon it will have to go into the [train_pipelines](https://github.com/alteryx/evalml/pull/1913/files#diff-8b4df814812e6f5df24c600bc7f429b036e0e2f78613bfbdcdf0ed3ab86bffd6R912) method that's going in #1913 . 

I don't have a clear alternative yet 😅  so I don't think this is something we need to change now. Maybe the engine can handle this responsibility and we can consider this in #1914

**Review by freddyaboulton — APPROVED** [review]
@bchen1116 I think this looks good! The one thing I want your input on is the order of sampling and creating a threshold tuning split in `_find_best_pipeline`.

File: evalml/automl/utils.py [hunk 8]
```diff
@@ -63,7 +64,7 @@ def make_data_splitter(X, y, problem_type, problem_configuration=None, n_splits=
     if problem_type == ProblemTypes.REGRESSION:
         return KFold(n_splits=n_splits, random_state=random_seed, shuffle=shuffle)
     elif problem_type in [ProblemTypes.BINARY, ProblemTypes.MULTICLASS]:
-        return StratifiedKFold(n_splits=n_splits, random_state=random_seed, shuffle=shuffle)
+        return BalancedClassificationDataCVSplit(n_splits=n_splits, random_seed=random_seed, shuffle=shuffle)
```
Comment (dsherry) [inline evalml/automl/utils.py:67]: 👍

File: evalml/tests/automl_tests/test_automl.py [hunk 9]
```diff
@@ -288,7 +292,7 @@ def _dummy_callback(param1, param2, param3):
         'Allowed Pipelines': [],
         'Patience': search_params['patience'],
         'Tolerance': search_params['tolerance'],
-        'Data Splitting': 'StratifiedKFold(n_splits=5, random_state=None, shuffle=False)',
+        'Data Splitting': ('BalancedClassificationDataCVSplit(balanced_ratio=None,', 'n_splits=5, random_seed=0'),
```
[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/tests/automl_tests/test_automl.py:295]: @freddyaboulton In this case, we are checking that each value in the tuple appears in the string, so `'BalancedClassificationDataCVSplit(balanced_ratio=None,'` and `'n_splits=5, random_seed=0'` are separately checked. We do this check because there's some weird spacing that happens in the output here, I think due to the formatting that these params are output in. I wasn't able to capture the spacing properly while making the string understandable, so I worked with this instead.

File: evalml/automl/utils.py [hunk 10]
```diff
@@ -63,7 +64,7 @@ def make_data_splitter(X, y, problem_type, problem_configuration=None, n_splits=
     if problem_type == ProblemTypes.REGRESSION:
```
[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/automl/utils.py:64]: Hmmm, that's a good question. It wasn't something I was going to add in this iteration, but it might be good to add? It's untested in terms of perf tests, however. @dsherry what do you think?

**Comment by dsherry** [comment]
@bchen1116 I left a few comments on the perf test doc, feel free to send me an invite if you wanna chat about any of that! Will review code shortly

File: evalml/automl/automl_search.py [hunk 11]
```diff
@@ -600,6 +600,10 @@ def _find_best_pipeline(self):
...
+                    train_indices = self.data_splitter.transform_sample(X_train, y_train)
```
[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/automl/automl_search.py:604]: Yep, that's a good call!

File: evalml/automl/automl_search.py (hunk 7 above)
Comment (bchen1116) [reply evalml/automl/automl_search.py:603]: @freddyaboulton Right, I agree, it does seem to be in too many places. Potentially if we move training the best pipeline to the engine, we could reduce the number of places this appears in. As for the `hasattr` call, we could also check the subclass name, if that's cleaner/preferred?

File: evalml/tests/automl_tests/test_automl.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply evalml/tests/automl_tests/test_automl.py:295]: Got it! Thanks for explaining that!

File: evalml/automl/automl_search.py (hunk 7 above)
Comment (freddyaboulton) [reply evalml/automl/automl_search.py]: @bchen1116 I think we can leave it as is for now!

**Review by dsherry — DISMISSED** [review]
@bchen1116 good stuff! This is looking nice so far.

Main points to change:
* I agree with @freddyaboulton that we should not break our current large data support, and that this PR should use the TV splitter instead of CV when appropriate. I left a suggestion for the control flow there.
* I left a suggestion for how to break up the unit tests a bit to make it easier for us to read and modify them in the future, and to make sure we have coverage of the math (which it appears you do!)
* A few impl suggestions
* Docstrings on the two splitter classes

File: evalml/preprocessing/data_splitters/balanced_classification_sampler.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+            y (pd.Series): Target data passed in
+
+        Returns:
+            (dict): dictionary with undersample target class as key, and number of samples to remove as the value.
+                If we don't need to resample, returns empty dictionary.
+        """
+        counts = y.value_counts()
+        normalized_counts = y.value_counts(normalize=True)
+        class_ratios = normalized_counts.copy().values
+        class_ratios /= class_ratios[-1]
```
[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/preprocessing/data_splitters/balanced_classification_sampler.py:54]: Mmm, good call. I changed to min, but this does work!

**Review by ParthivNaresh — APPROVED** [review]
Great work, the new parametrized tests look excellent!

File: evalml/preprocessing/data_splitters/balanced_classification_splitter.py [hunk 13]
```diff
@@ -24,69 +29,44 @@ def __init__(self, balanced_ratio=4, min_samples=100, min_percentage=0.1, random
                 If class percentage and min_samples are not met, treat this as severely imbalanced, and we will not resample the data.
                 Must be between 0 and 0.5, inclusive. Defaults to 0.1.
 
+            test_size (float): The size of the test split. If none provided, this is set to 0.25. Defaults to None.
```
[the remark this answers is not in the document]
Comment (bchen1116) [reply evalml/preprocessing/data_splitters/balanced_classification_splitter.py:32]: Mmm, I guess it can default to 0.25. Was a little confusing haha
